[filter_single] Added tensor_filter_single prototype

kparichay opened the pull request:
Added prototype for tensor_filter_single
tensor_filter_single is going to form the backend for singleshot api
to minimize the latency. This is made to be independent of GStreamer,
supporting all the existing tensor_filter plugins

Working on #1547 

**Self evaluation:**
1. Build test: [x]Passed [ ]Failed [ ]Skipped
2. Run test: [ ]Passed [ ]Failed [*]Skipped

Signed-off-by: Parichay Kapoor <pk.kapoor@samsung.com>

myungjoo commented:
1. Why do we need ```gst/nnstreamer/tensor_filter/tensor_filter_single.h```? tensor_filter_single.c should not export its own interface; its interface should be exactly same with the conventional single-shot API.

2. This should not be built as another "tensor filter subplugin", but as an independent shared library or static library.

kparichay commented:
> 1. Why do we need `gst/nnstreamer/tensor_filter/tensor_filter_single.h`? tensor_filter_single.c should not export its own interface; its interface should be exactly same with the conventional single-shot API.

tensor_filter_single.c implements tensor_filter without gst support. There will be single-shot c-API new implementation which will have same interface as conventional single-shot API

> 2. This should not be built as another "tensor filter subplugin", but as an independent shared library or static library.

Updated. Moved to `api/capi`. An independent static library for `tensor_filter_single`, which is linked with `capi`.

myungjoo commented on api/capi/include/tensor_filter_single.h:
```diff
@@ -0,0 +1,84 @@
+#include <glib-object.h>
+
+#include "nnstreamer_subplugin.h"
+#include "nnstreamer_plugin_api_filter.h"
```
They are not in your local directory, try to use <nnstreamer/*.h> and drop nnsinc.

kparichay replied:
updated.

myungjoo commented:
looks good in overall.

kparichay commented:
V2:
Updated `#include` paths to format `<nnstreamer/....>`
Removed `nnsinc` include_directories from meson.build
Added more vmethods for tensor_filter single class

myungjoo commented on api/capi/meson.build:
```diff
@@ -18,6 +18,21 @@ if meson.project_name() != 'nnstreamer'
+nninc = include_directories('../../gst')
+
+# Create dependency for tensor_filter_single for single-shot capi
+tensor_filter_single_lib = static_library ('tensor_filter_single_lib',
```
Any reason to be a static library?

kparichay replied:
Updated to shared library.

kparichay commented:
V3:
`tensor_filter_single` is a shared_library
Added `gobject-2.0` as a dependency

V2:
Updated #include paths to format <nnstreamer/....>
Removed nnsinc include_directories from meson.build
Added more vmethods for tensor_filter single class
